### PR TITLE
Component side tray

### PR DESCRIPTION
**Pull request by chelshaw**
**SUMMARY**
Add a sidetray component that calls closeCallback when:
- You click outside the sidetray
- You click the `esc` key

Has props for header and footer nodes, and content between them automatically flexGrows

### CODE REVIEW

File: src/components/SideTray/index.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        </div>
+      </div>
+      {clickOffBackdrop}
+    </div>
+  );
+};
+
+SideTray.defaultProps = {
+  visible: false,
+  modifiers: [],
```
[the remark this answers is not in the document]
Comment (chelshaw) [reply src/components/SideTray/index.js]: oh yeah. I removed it from props but not defaults :D

File: src/components/SideTray/style.css [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+.main {
+  display: flex;
+  flex-direction: column;
+  justify-content: space-evenly;
+  overflow: hidden;
+  height: 100%;
+}
+
+.content {
+  flex-grow: 1;
```
Comment (pixelbandito) [inline src/components/SideTray/style.css]: Any time something resizes with flex, you need `min-width: 0` / `min-height: 0` (depending on the flow).

File: src/components/SideTray/style.css [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  right: 0;
+  bottom: 0;
+  left: 0;
+  height: 100vh;
+  width: 100vw;
+  z-index: var(--rvr-off-canvas-drawer-z-index) - 1;
+  background-color: rgba(0, 0, 0, 0.3);
+  overflow: auto;
+  opacity: 10;
+  transition: opacity 150ms var(--rvr-easeInOutQuad);
```
Comment (pixelbandito) [inline src/components/SideTray/style.css]: The variables for timing functions are a good start. Wonder if we could come up with a generic pattern to inform when to use what? (I just posted about [this](https://uxdesign.cc/the-ultimate-guide-to-proper-use-of-animation-in-ux-10bd98614fa9) in eng-guild-front-end)

I was thinking something like

Timing function variables
* timing-fade
* timing-move-in
* timing-move-out

Durations
* slow
* medium
* fast

For this element, since it's pretty big, we can do "slow timing-fade" for the backdrop and "slow timing-move-in"/"slow timing-move-out" for the actual tray. We'd want to go back and use a "fast" duration for buttons, and we could even come back later and adjust speeds based on media queries.

File: src/components/SideTray/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# <SideTray>
+
+Add your documentation for the `SideTray` component here.
```
Comment (pixelbandito) [inline src/components/SideTray/README.md]: This would be nice.

File: example/yarn.lock [hunk 5]
```diff
@@ -217,9 +217,9 @@ ajv@^5.0.0, ajv@^5.1.5, ajv@^5.2.0:
     json-schema-traverse "^0.3.0"
 
 ajv@^6.0.1, ajv@^6.5.5:
-  version "6.10.1"
-  resolved "https://trendkite.jfrog.io/trendkite/api/npm/libs-frontend/ajv/-/ajv-6.10.1.tgz#ebf8d3af22552df9dd049bfbe50cc2390e823593"
-  integrity sha1-6/jTryJVLfndBJv75QzCOQ6CNZM=
+  version "6.10.2"
```
Comment (pixelbandito) [inline example/yarn.lock]: Do you have time to add a .gitattributes file for this?
```
yarn.lock             -diff
example/yarn.lock     -diff
```

File: src/components/SideTray/story.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  );
+}
+
+storiesOf('Star System/SideTray', module)
+  .addParameters({
+    readme: {
+      sidebar: Readme,
+    },
+  })
+  .add('Overview', () => <SideTrayTest />, {
```
Comment (pixelbandito) [inline src/components/SideTray/story.js]: This is the "Example"
The "Overview" should be the bare component with only knobs and actions

**Comment by pixelbandito** [comment]
I committed a change to add a SideTray to the example page. Would you verify it doesn't break your subsequent changes?

Also, I tested in IE11 and Edge - looking good!!

File: src/components/SideTray/style.css (hunk 3 above)
Comment (chelshaw) [reply src/components/SideTray/style.css]: I'm not gonna do this at the moment; i'm not even convinced 150ms is a quantity worth transitioning over

File: src/components/SideTray/Body/index.js [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import style from './style.css';
+
+const Body = ({ children, ...passedProps }) => {
+  return (
+    <div className={style.SideTray__Body} {...passedProps}>
```
Comment (pixelbandito) [inline src/components/SideTray/Body/index.js]: Style name could just be `Body`, since css modules should be taking care of uniqueness.

And you should probably concat `style.Body` to whatever `props.className` the consumer passed in.

File: src/components/SideTray/Body/index.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+      {children}
```
Comment (pixelbandito) [inline src/components/SideTray/Body/index.js]: You don't need to enumerate children. It's just a prop, so let `passedProps` handle it.

File: src/components/SideTray/Footer/index.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import style from './style.css';
+
+const Footer = ({ children, ...passedProps }) => {
+  return (
+    <div className={style.SideTray__Footer} {...passedProps}>
```
Comment (pixelbandito) [inline src/components/SideTray/Footer/index.js]: Same comments as body.

File: src/components/SideTray/Header/index.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import style from './style.css';
+
+const Header = ({ children, ...passedProps }) => {
```
Comment (pixelbandito) [inline src/components/SideTray/Header/index.js]: 🔁

File: src/stories.js [hunk 11]
```diff
@@ -1,5 +1,6 @@
 import './shared/colors.css';
 import './shared/sizing.css';
+import './shared/buttons.css';
```
Comment (pixelbandito) [inline src/stories.js]: I'm not 100%, but we might be able to drop these since we're doing the CSS loader config to make the variables available.

Comment (chelshaw) [reply src/stories.js]: when this was missing my buttons in storybook weren't working

File: src/components/SideTray/Body/index.js (hunk 7 above)
Comment (mdespuits) [reply src/components/SideTray/Body/index.js]: If you take a look at the `Media.jsx` in tk-ui I have a pattern that does this exact thing if that helps

File: src/stories.js (hunk 11 above)
Comment (mdespuits) [reply src/stories.js]: This was not working in Storybook because the loader changes were made in the rollup config, afaik

File: src/components/SideTray/style.css [hunk 12]
```diff
@@ -1,5 +1,5 @@
-.sideTray {
-  width: var(--rvr-side-tray-width);
```
Comment (mdespuits) [inline src/components/SideTray/style.css]: Why are we losing the variable here?

**Comment by pixelbandito** [comment]
Good catch on the imports in stories.i think that's because storybook used a different webpack config and therefore different css loader module options, 

Importing shared css directly for now in stories is great.

File: src/components/SideTray/style.css (hunk 12 above)
Comment (pixelbandito) [reply src/components/SideTray/style.css]: Hey @mdespuits,

I suggested this. I figured that:
1. The sidetray width isn't intended for re-use among components. 
2. There's nothing magical about 400 px in our sizing system.
3. It's not part of the theme.

I could see us standardizing on some sizes for modals / popovers / etc. in the future, but for now I thought this was cleaner than having a block of sidebar-specific properties in our global variables.

File: src/stories.js (hunk 11 above)
Comment (pixelbandito) [reply src/stories.js]: Yup, so this is good.

**Comment by pixelbandito** [comment]
I'm drunk with power and seeing if I can dismiss @mdespuits review, but only because I already verified that @chelshaw made the changes requested.
